pkg/planner/core/task.go

```diff
@@ -1000,6 +1004,50 @@ func (p *PhysicalTopN) Attach2Task(tasks ...base.Task) base.Task {
 	return attachPlan2Task(p, rootTask)
 }
 
+var annIndexFnNameToMetric = map[string]tipb.VectorDistanceMetric{
+	ast.VecL1Distance:     tipb.VectorDistanceMetric_L1,
+	ast.VecL2Distance:     tipb.VectorDistanceMetric_L2,
+	ast.VecCosineDistance: tipb.VectorDistanceMetric_COSINE,
+
+	// Note: IP is not supported yet. Currently we will throw errors when building IP index.
+	ast.VecNegativeInnerProduct: tipb.VectorDistanceMetric_INNER_PRODUCT,
+}
+
+func fixTopNForANNIndex(p *PhysicalTopN) bool {
+	ts, ok := p.Children()[0].(*PhysicalTableScan)
+	if !ok || ts.AnnIndexExtra == nil {
+		return true
+	}
+	vecColID := ts.AnnIndexExtra.ColumnId
+	if len(p.ByItems) != 1 {
+		return false
+	}
+	if p.ByItems[0].Desc {
+		// Currently all vector search indexes must be used for ASC order.
+		// DESC order can be only applied to INNER_PRODUCT, which is not
+		// supported yet.
+		return false
+	}
+	vs, err := expression.ExtractVectorSearch(p.ByItems[0].Expr)
+	if err != nil || vs == nil {
+		return false
+	}
+	disctanceMetricPB, ok := annIndexFnNameToMetric[vs.DistanceFnName.L]
+	if !ok {
+		return false
+	}
+	if disctanceMetricPB != ts.AnnIndexExtra.DistanceMetric {
+		return false
+	}
+	if vs.Column.ID != vecColID {
+		return false
+	}
+	ts.AnnIndexExtra.RefVecF32 = vs.Vec.SerializeTo(nil)
+	ts.AnnIndexExtra.TopK = uint32(p.Count)
```

Comment:
better add some comments for this function

we could only see: that it checks some elements in topN's by item, then fills some fields in ts. what is it used for? another projection injection logic is not seen here?